spec.html

```diff
@@ -37634,12 +37549,86 @@ <h1>
+        </emu-note>
+        <emu-note>
+          <p>Because non-existent property values always compare greater than *undefined* property values, and *undefined* always compares greater than any other value, *undefined* property values always sort to the end of the result, followed by non-existent property values.</p>
+        </emu-note>
+        <emu-note>
+          <p>The `sort` function is intentionally generic; it does not require that its *this* value be an Array. Therefore, it can be transferred to other kinds of objects for use as a method.</p>
+        </emu-note>
+
+        <emu-clause id="sec-sortindexedproperties" type="abstract operation">
+          <h1>
+            SortIndexedProperties (
+              _obj_: an Object,
+              _len_: a non-negative integer,
+              _SortCompare_: an Abstract Closure with two parameters,
+            )
+          </h1>
+          <dl class="header">
+          </dl>
+          <emu-alg>
+            1. Let _items_ be a new empty List.
+            1. Let _k_ be 0.
+            1. Repeat, while _k_ &lt; _len_,
+              1. Let _Pk_ be ! ToString(𝔽(_k_)).
+              1. Let _kPresent_ be ? HasProperty(_obj_, _Pk_).
+              1. If _kPresent_ is *true*, then
+                1. Let _kValue_ be ? Get(_obj_, _Pk_).
+                1. Append _kValue_ to _items_.
+              1. Set _k_ to _k_ + 1.
+            1. Let _itemCount_ be the number of elements in _items_.
+            1. [id="step-array-sort"] Sort _items_ using an implementation-defined sequence of calls to _SortCompare_. If any such call returns an abrupt completion, stop before performing any further calls to _SortCompare_ or steps in this algorithm and return that completion.
+            1. Let _j_ be 0.
+            1. Repeat, while _j_ &lt; _itemCount_,
+              1. Perform ? Set(_obj_, ! ToString(𝔽(_j_)), _items_[_j_], *true*).
+              1. Set _j_ to _j_ + 1.
+            1. Repeat, while _j_ &lt; _len_,
+              1. Perform ? DeletePropertyOrThrow(_obj_, ! ToString(𝔽(_j_))).
+              1. Set _j_ to _j_ + 1.
+            1. Return _obj_.
           </emu-alg>
+          <p>The <dfn id="sort-order">sort order</dfn> is the ordering of _items_ after completion of step <emu-xref href="#step-array-sort"></emu-xref> of the algorithm above. The sort order is implementation-defined if _SortCompare_ is not a consistent comparator for the elements of _items_ (see below). The sort order is also implementation-defined if SortIndexedProperties is invoked by <emu-xref href="#sec-array.prototype.sort">Array.prototype.sort</emu-xref>, _comparefn_ is *undefined*, and all applications of ToString, to any specific value passed as an argument to _SortCompare_, do not produce the same result.</p>
```

Comment:
Why the reference to `Array.prototype.sort`?

It's true that the only other caller of `SortIndexedProperties` is TA.prototype.sort, and in that case the elements are all primitives and so ToString will necessarily be consistent, so this requirement is redundant in that case. But it still seems odd to explicitly carve out `Array.prototype.sort` here; it suggests to the reader that this requirement is _not required_ in the TypedArray case, rather than being redundant there.

Reply:
> It's true that the only other caller of `SortIndexedProperties` is TA.prototype.sort, and in that case the elements are all primitives and so ToString will necessarily be consistent,

TA.p.sort's _SortCompare_ doesn't call ToString. I think it might confuse the reader to suggest that it does. I suppose we could have a Note explaining why there isn't a corresponding imp-def criterion for TA.

Reply:
The thing is, specifically referring (and linking) to A.p.sort gives a context for the rest of the sentence, so there's a fair chance that the reader will understand that `_comparefn_` refers to the A.p.sort parameter.

If we delete that reference, then it would look more like `_comparefn_` refers to an alias in SortIndexedProperties, which might puzzle readers. Myself, I'd be inclined to guess it was a copy/paste error.

Now, you could address that problem by saying something like "The sort order is also implementation-defined if the function that calls SortIndexedProperties has a `_comparefn_` parameter, and the value of that parameter is `*undefined*`, and [etc]." That at least would set up the context and look intentional.

But to me, it seems like unnecessary generalization: we know that the sentence only applies A.p.sort -- why not just say so? Are you thinking that there might be other callers in the future, and they will automatically be handled correctly by the general text? I'm not so sure.

But I'm willing to change it to something like that if you still think it's an improvement.

Reply:
> The thing is, specifically referring (and linking) to A.p.sort gives a context for the rest of the sentence, so there's a fair chance that the reader will understand that `_comparefn_` refers to the A.p.sort parameter.

Hm, that's fair, but I think that can be conveyed more clearly with a different phrasing. Perhaps just reordering this sentence a bit? As in

```suggestion
          <p>The <dfn id="sort-order">sort order</dfn> is the ordering of _items_ after completion of step <emu-xref href="#step-array-sort"></emu-xref> of the algorithm above. The sort order is implementation-defined if _SortCompare_ is not a consistent comparator for the elements of _items_ (see below). When SortIndexedProperties is invoked by <emu-xref href="#sec-array.prototype.sort">Array.prototype.sort</emu-xref>, the sort order is also implementation-defined if _comparefn_ is *undefined* and all applications of ToString, to any specific value passed as an argument to _SortCompare_, do not produce the same result.</p>

```
---

> Are you thinking that there might be other callers in the future, and they will automatically be handled correctly by the general text?

Well, there is [a currently proposed new caller](https://tc39.es/proposal-change-array-by-copy/#sec-array.prototype.toSorted), but that wasn't actually the motivation. It really was just what I put above: the current phrasing to me suggests that it is _specifically exempting_ TA.prototype.sort, rather than (as is actually the case) simply not being relevant to that algorithm.

Reply:
I'm happy with the wording now. I don't necessarily much mind mentioning A.p.sort, and I agree it's kind of tricky to generalize in a way which is clearly not normative.

Reply:
> > > Are you thinking that there might be other callers in the future, and they will automatically be handled correctly by the general text?
> > 
> > 
> > Well, there is [a currently proposed new caller](https://tc39.es/proposal-change-array-by-copy/#sec-array.prototype.toSorted),
> 
> It has different semantics from what SortIndexedProperties currently provides, though.

Could `sortindexedproperties` return the internal sorted spec list? So the mutating calls to `Set` are kept in the methods algos? Thinking that might allow [toSorted](https://tc39.es/proposal-change-array-by-copy/#sec-array.prototype.toSorted) to share more of key notes on when sort is implementation defined?

Reply:
I think the spec changes for Change-array-by-copy would likely add another level of indirection to extract out the actual sort on step 5 to a new AO that takes and returns a spec list, and the notes for this section would move there too.
This would allow it to be shared in both the mutating/immutable and hole aware/ignoring methods.

Reply:
> ... extract out the actual sort on step 5 to a new AO that takes and returns a spec list, and the notes for this section would move there too. This would allow it to be shared in both the mutating/immutable and hole aware/ignoring methods.

Yeah, I made myself a note about factoring out something like `HostSortList(_items_, _SortCompare_)`.  I didn't think it was necessary for this PR, but I'm willing if the editors want it.

Reply:
If it will make the change array by copy proposal easier to review, it'd be nice to get it in before the March meeting.

Reply:
Doesn't need to be done in this PR, in my opinion.

Reply:
(Generally I'm of the opinion that changes should be well-motivated in their own right; if we would only be making a change for the sake of a later PR, we should just make that change in that PR.)

Reply:
I'm happy either way, the change being in this PR or as part of the proposal. Suggestion was only to reduce the frequency of spec chance in this area 🙂

Reply:
I think that change is more appropriate as part of the proposal.